libfeed/feed-store-provider.c

```diff
@@ -0,0 +1,1239 @@
+                                          const gchar          *source_name,
+                                          ModelFromResultFunc   marshal_func,
+                                          gpointer              marshal_data,
+                                          GCancellable         *cancellable,
+                                          GError              **error)
+{
+  g_autoptr(GPtrArray) model_data_array = NULL;
+  g_autoptr(GVariant) result = g_dbus_proxy_call_sync (proxy,
+                                                       method_name,
+                                                       NULL,
+                                                       G_DBUS_CALL_FLAGS_NONE,
+                                                       -1,
+                                                       cancellable,
+                                                       error);
+  g_autoptr(GVariant) model_variant = NULL;
+  g_autoptr(GHashTable) model_ht = NULL;
+
+  if (result == NULL)
+    return NULL;
+
+  g_variant_get (result, "(@a{ss})", &model_variant);
+  model_ht = hashtable_from_variant (model_variant);
+
+  /* Now that we have the model, we can marshal it into a
+   * GObject */
+  return marshal_func (model_ht, marshal_data);
+}
+
+/* Quite surprising that I could not find an alternative to this in Gio */
+static gchar *
+remove_uri_prefix (const gchar  *uri,
+                   GError      **error)
+{
+  g_autoptr(GError) local_error = NULL;
+  g_autoptr(GRegex) regex = g_regex_new ("[a-z]+:\\/+(\\w+)", 0, 0, &local_error);
+  g_autoptr(GMatchInfo) match_info = NULL;
+
+  if (regex == NULL)
+    {
+      g_error ("Failed to compile regex: %s", local_error->message);
+      return NULL;
+    }
+
+  if (g_regex_match (regex, uri, 0, &match_info))
+    {
+      return g_match_info_fetch (match_info, 1);
+    }
+
+  g_set_error (error,
+               G_IO_ERROR,
+               G_IO_ERROR_FAILED,
+               "Failed to parse URI %s", uri);
+  return NULL;
+}
+
+static GInputStream *
+find_thumbnail_stream_in_shards (const gchar * const  *shards_strv,
+                                 const gchar          *thumbnail_uri)
+{
+  g_autoptr(GError) local_error = NULL;
+  const gchar * const *iter = shards_strv;
+  g_autofree gchar *normalized = remove_uri_prefix (thumbnail_uri, &local_error);
+
+  if (normalized == NULL)
+    {
+      g_message ("Can't load thumbnail: %s", local_error->message);
+      return NULL;
+    }
+
+  for (; *iter != NULL; ++iter) {
+    g_autoptr(EosShardShardFile) shard_file = g_object_new (EOS_SHARD_TYPE_SHARD_FILE,
+                                                            "path", *iter,
+                                                            NULL);
+    g_autoptr(EosShardRecord) record = NULL;
+    g_autoptr(GError) local_error = NULL;
+
+    /* XXX: This should probably be done asynchronously if possible */
+    if (!g_initable_init (G_INITABLE (shard_file), NULL, &local_error))
```

Comment:
It wasn't in the JS code either, just didn't want to do it here to avoid extra churn. We do it async in the companion app and it works fine though.